File: go/reviewability_patterns.go
//go:build reviewability

// Package benchmarks contains exact template patterns for benchmark testing.
// Each function contains the EXACT OriginalCode from templates.go.
//
// Reviewability fixtures carry a dedicated build tag: their stub types (Order,
// User) reuse names already defined in patterns.go, so they must not compile
// under the shared "patterns" tag.
package benchmarks

import "errors"

type reviewOrderStatus string

const StatusPaid reviewOrderStatus = "paid"

type Order struct {
	ID     string
	Status reviewOrderStatus
}

type User struct {
	approver bool
}

func (u User) CanApproveOrders() bool { return u.approver }

var ErrForbidden = errors.New("forbidden")

func markApproved(id string) error    { return nil }
func enqueueShipment(id string) error { return nil }

func canApproveOrder(order Order, user User) bool {
	return order.Status == StatusPaid && user.CanApproveOrders()
}

type Invoice struct {
	CustomerTier  string
	SubtotalCents int
	DiscountCents int
}

type DiscountPolicy struct{}

func (DiscountPolicy) DiscountFor(tier string, subtotalCents int) int { return 0 }

// go-reviewability-duplicated-policy-hard - exact multi-line
func canFinalizeOrder(order Order, user User) bool {
	return order.Status == StatusPaid && user.CanApproveOrders()
}

func approveOrder(order Order, user User) error {
	if !canFinalizeOrder(order, user) {
		return ErrForbidden
	}
	return markApproved(order.ID)
}

func shipOrder(order Order, user User) error {
	if !canFinalizeOrder(order, user) {
		return ErrForbidden
	}
	return enqueueShipment(order.ID)
}

// go-reviewability-ownership-boundary-hard - exact multi-line
func ApplyInvoiceDiscount(invoice *Invoice, policy DiscountPolicy) {
	invoice.DiscountCents = policy.DiscountFor(invoice.CustomerTier, invoice.SubtotalCents)
}

// go-fp-reviewability-central-policy - exact multi-line
func approveOrderWithSharedPolicy(order Order, user User) error {
	if !canApproveOrder(order, user) {
		return ErrForbidden
	}
	return markApproved(order.ID)
}

File: python/reviewability_patterns.py
"""Benchmark patterns for Python reviewability (future-maintainer change risk).

Each block contains the EXACT OriginalCode from templates.go.
"""


class Config:
    pass


class Order:
    status: str
    region: str


class User:
    allowed_regions: list


def issue_refund(order: "Order") -> None:
    ...


# py-reviewability-temporal-coupling-medium - exact multi-line
class SearchClient:
    @classmethod
    def connect(cls, config: Config) -> "SearchClient":
        client = cls(config)
        client._open()
        return client

    def __init__(self, config: Config) -> None:
        self.config = config

    def _open(self) -> None:
        ...

    def index(self, document: object) -> None:
        ...

    def close(self) -> None:
        ...


# py-reviewability-divergent-policy-hard - exact multi-line
def can_refund(order: Order, user: User) -> bool:
    return order.status == "paid" and order.region in user.allowed_regions

def refund_order(order: Order, user: User) -> None:
    if not can_refund(order, user):
        raise PermissionError()
    issue_refund(order)


config = Config()
document = object()

# py-fp-reviewability-context-manager - exact multi-line
client = SearchClient.connect(config)
client.index(document)
client.close()
